src/modules/mavlink/mavlink_main.cpp

```diff
@@ -784,40 +784,55 @@ Mavlink::send_packet()
 {
 	int ret = -1;
 
-#if defined(CONFIG_NET) || defined(__PX4_POSIX)
+	// only transmit after we've received messages
+	if (!should_transmit()) {
+		_buf_len = 0;
+	}
 
-	/* Only send packets if there is something in the buffer. */
-	if (_network_buf_len == 0) {
+	// Only send packets if there is something in the buffer
+	if (_buf_len == 0) {
 		pthread_mutex_unlock(&_send_mutex);
```

Comment:
Why is there an `unlock` inside this function and no `lock`? That's odd, right?

Reply:
Hopefully explained in the main comment. 

    Mavlink::send_packet() = MAVLINK_END_UART_SEND

Perhaps we should rename it...

Reply:
Ok